### PR TITLE
store weak pointer to ucc work

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**

### CODE REVIEW

File: src/torch_ucc.cpp [hunk 1]
```diff
@@ -133,65 +128,63 @@ ProcessGroupUCC::WorkUCC::~WorkUCC() {
...
 }
 
 bool ProcessGroupUCC::WorkUCC::isSuccess() const {
-  return (status_ >= 0);
+  return !exception() && entry_->status_ == 0;
 }
 
 bool ProcessGroupUCC::WorkUCC::wait(std::chrono::milliseconds /* unused */) {
 #ifdef USE_CUDA
   if (fence && !torch_ucc_config.blocking_wait[(int)opType_]) {
```
Comment (kingchc) [inline src/torch_ucc.cpp:156]: not relate to this PR, just a question of a corner case comes to mind now - what if a request is progressed and completed before main thread calls wait (e.g., small collective)? it seems no point to block the stream anyhow here in such scenario and we may save a few micro/nano-seconds?

**Review by kingchc — COMMENTED** [review]
@Sergei-Lebedev - thanks for this PR! I have tested the updated one and can verify it fixes the issue we encountered. I have a few comments here, please take a look at it and let me know if it makes sense.

File: src/torch_ucc.cpp [hunk 2]
```diff
@@ -133,65 +128,63 @@ ProcessGroupUCC::WorkUCC::~WorkUCC() {
...
     // block user stream
+    setAndThrowException();
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/torch_ucc.cpp:158]: No, main thread will get exception ptr only when progress thread set it. If throw exception directly from progress thread nobody will be able to catch it

File: src/torch_ucc.cpp [hunk 3]
```diff
@@ -582,49 +572,57 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupUCC::collective_post(
...
+        }
+      }
+      cuda_ev->record(at::cuda::getCurrentCUDAStream(dev.index()));
+      cuda_ev->block(*stream);
+      if (torch_ucc_config.use_future) {
+        c10::cuda::CUDAMultiStreamGuard streamGuard(*stream);
+        std::vector<c10::Device> devList{dev};
+        work->future_ = c10::make_intrusive<at::ivalue::Future>(
+            c10::ListType::create(c10::TensorType::get()), devList);
       }
```
[the remark this answers is not in the document]
Comment (Sergei-Lebedev) [reply src/torch_ucc.cpp:609]: maybe remove torch_ucc_config.use_future completely? Do we have use case to disable futures?

File: src/torch_ucc.cpp (hunk 1 above)
Comment (Sergei-Lebedev) [reply src/torch_ucc.cpp:156]: I agree, moreover we can check that collective completes immediately in enqueue_collective or enqueue_cuda_collective and do not even bother progress thread in that case. Maybe create a separate PR for this change?

**Comment by Sergei-Lebedev** [comment]
@kingchc  also I think we can remove cuda event pool now since progress thread doesn't hold pointer to it anymore

**Comment by kingchc** [comment]
close this PR as we merged https://github.com/facebookresearch/torch_ucc/pull/28, where we cherry-picked this PR onto the latest master. (cc @Sergei-Lebedev)

**Pull request closed without merging**